go/store/nbs/journal_writer.go

```diff
@@ -187,36 +188,67 @@ func (wr *journalWriter) bootstrapJournal(ctx context.Context) (last hash.Hash,
 		if info, err = wr.index.Stat(); err != nil {
 			return hash.Hash{}, err
 		}
+
+		// initialize range index with enough capacity to
+		// avoid rehashing during bootstrapping
+		cnt := estimateRangeCount(info)
+		wr.ranges.cached = swiss.NewMap[addr16, Range](cnt)
+
+		eg, ectx := errgroup.WithContext(ctx)
+		ch := make(chan []lookup, 4)
+
 		// process the indexed portion of the journal
-		err = processIndexRecords(ctx, wr.index, info.Size(), func(o int64, r indexRec) (err error) {
-			switch r.kind {
-			case tableIndexRecKind:
-				// |r.end| is expected to point to a root hash record in |wr.journal|
-				// containing a hash equal to |r.lastRoot|, validate this here
-				var h hash.Hash
-				if h, err = peekRootHashAt(wr.journal, int64(r.end)); err != nil {
-					return err
-				} else if h != r.lastRoot {
-					return fmt.Errorf("invalid index record hash (%s != %s)", h.String(), r.lastRoot.String())
+		eg.Go(func() error {
+			defer close(ch)
+			return processIndexRecords(ectx, wr.index, info.Size(), func(o int64, r indexRec) (err error) {
+				switch r.kind {
+				case tableIndexRecKind:
+					// |r.end| is expected to point to a root hash record in |wr.journal|
+					// containing a hash equal to |r.lastRoot|, validate this here
+					var h hash.Hash
+					if h, err = peekRootHashAt(wr.journal, int64(r.end)); err != nil {
+						return err
+					} else if h != r.lastRoot {
+						return fmt.Errorf("invalid index record hash (%s != %s)", h.String(), r.lastRoot.String())
+					}
+					select {
+					case <-ectx.Done():
+						return ectx.Err()
+					case ch <- deserializeLookups(r.payload):
+						// record a high-water-mark for the indexed portion of the journal
+						wr.indexed = int64(r.end)
+					}
+					// todo: uncompressed size
+				default:
+					return fmt.Errorf("unknown index record kind (%d)", r.kind)
 				}
-				// populate range hashmap
-				for _, l := range deserializeLookups(r.payload) {
-					wr.ranges.put(l.a, l.r)
+				return nil
+			})
+		})
+		// populate range hashmap
+		eg.Go(func() error {
+			for {
+				select {
+				case <-ectx.Done():
+					return nil
+				case ll, ok := <-ch:
+					if !ok {
+						return nil
+					}
+					for _, l := range ll {
+						wr.ranges.putCached(l.a, l.r)
+					}
 				}
-				// record a high-water-mark for the indexed portion of the journal
-				wr.indexed = int64(r.end)
-				wr.ranges = wr.ranges.flatten()
-				// todo: uncompressed size
-			default:
-				return fmt.Errorf("unknown index record kind (%d)", r.kind)
 			}
-			return nil
 		})
+
+		err = eg.Wait()
 		if err != nil {
-			// todo: issue warning on corrupt index recovery
```

Comment:
failing to process `journal.idx` was considered a recoverable error, so we were (always) silently dropping the journal index